lib/protobuf/field/message_field.rb

```diff
@@ -35,8 +35,8 @@ def coerce!(value)
           nil
         elsif value.is_a?(type_class)
           value
-        elsif value.respond_to?(:to_proto)
-          value.to_proto
+        elsif value.respond_to?(:to_proto) && (proto_value = value.to_proto).is_a?(type_class)
```

Comment:
@liveh2o I think this patch is pretty reasonable as-is, I'm happy leaving it as @embark proposed?